solaris/src/evm.rs

```diff
@@ -114,9 +114,16 @@ impl Evm {
         self
     }
 
-    pub fn logs(&self, _filter: ::ethabi::TopicFilter) -> Vec<()> {
-        // TODO [ToDr] Add filter querying
-        self.logs.iter().map(|_| ()).collect()
+    pub fn logs<T: Into<Option<ethabi::TopicFilter>>>(
```

Comment:
IMHO it  would be perfect if the logs were decode immediately to expected format:
```
let deposit = contract.events().deposit();
let parsed_logs = logs(deposit.create_filter(), |log| deposit.parse_log(log))?;
```
This gives us more guarantees than just checking `len()` - it might be a false positive.

But a great start anyway!

Reply:
in my tests so far i've been asserting`len()` first, then look at individual events and parse them. no false positives this way.

on the API:

how about having `logs()` return an iterator? then one can simply `.filter(sol::adapt_filter(deposit.create_filter())` to filter by event and`.map(|log| deposit.parse_log(log)).unwrap()` to get usable log structs. that would keep things composable and flexible

Reply:
we could have a convenience function `event_logs(deposit)` that does both steps automatically for a given event

Reply:
events generated by ethabi derive all seem to have functions `parse_log` and `create_filter` with identical signatures (see here and below: https://gist.github.com/snd/8dad886c6acdf057a18e2c29b6255c48#file-foreign_bridge-rs-L84).

if we'd move these functions into a trait `EthabiEvent` (for example) then `evm.event_logs` could have the very convenient interface: `fn event_logs<T: EthabiEvent>(&self) -> SomeIteratorType<T>` and usage `deposit_events: Vec<Deposit> = evm.event_logs();`

Reply:
You can still have a trait with associated type for return value. See #6 for similar proposal of `EthabiFunction` trait - this would improve the API as well.

I'll have a look at required changes in ethabi this week.

Reply:
didn't see the comment in time and came up with a similar idea over here https://github.com/paritytech/sol-rs/issues/12#issuecomment-352680549

i'd love to help  implementing this change. let me know